src/aggregation/bucket/range.rs

```diff
@@ -0,0 +1,546 @@
+            to: range.end,
+        }
+    }
+}
+
+impl From<&RangeAggregationRange> for Range<f64> {
+    fn from(range: &RangeAggregationRange) -> Self {
+        range.from..range.to
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SegmentRangeAndBucketEntry {
+    range: Range<u64>,
+    bucket: SegmentRangeBucketEntry,
+}
+
+/// The collector puts values from the fast field into the correct buckets and does a conversion to
+/// the correct datatype.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SegmentRangeCollector {
+    /// The buckets containing the aggregation data.
+    buckets: Vec<SegmentRangeAndBucketEntry>,
+    field_type: Type,
+}
+
+impl SegmentRangeCollector {
+    pub fn into_intermediate_bucket_result(self) -> IntermediateBucketResult {
+        let field_type = self.field_type;
+
+        let buckets = self
+            .buckets
+            .into_iter()
+            .map(move |range_bucket| {
+                (
+                    range_to_key(&range_bucket.range, &field_type),
+                    range_bucket.bucket.into(),
+                )
+            })
+            .collect();
+
+        IntermediateBucketResult::Range(buckets)
+    }
+
+    pub(crate) fn from_req(
+        req: &RangeAggregation,
+        sub_aggregation: &AggregationsWithAccessor,
+        field_type: Type,
+    ) -> crate::Result<Self> {
+        // The range input on the request is f64.
+        // We need to convert to u64 ranges, because we read the values as u64.
+        // The mapping from the conversion is monotonic so ordering is preserved.
+        let buckets = extend_validate_ranges(&req.ranges, &field_type)?
+            .iter()
+            .map(|range| {
+                let to = if range.end == u64::MAX {
+                    None
+                } else {
+                    Some(f64_from_fastfield_u64(range.end, &field_type))
+                };
+                let from = if range.start == u64::MIN {
+                    None
+                } else {
+                    Some(f64_from_fastfield_u64(range.start, &field_type))
+                };
+                Ok(SegmentRangeAndBucketEntry {
+                    range: range.clone(),
+                    bucket: SegmentRangeBucketEntry {
+                        key: range_to_key(range, &field_type),
+                        doc_count: 0,
+                        sub_aggregation: SegmentAggregationResultsCollector::from_req(
+                            sub_aggregation,
+                        )?,
+                        from,
+                        to,
+                    },
+                })
+            })
+            .collect::<crate::Result<_>>()?;
+
+        Ok(SegmentRangeCollector {
+            buckets,
+            field_type,
+        })
+    }
+
+    #[inline]
+    pub(crate) fn collect_block(
+        &mut self,
+        doc: &[DocId],
+        bucket_with_accessor: &BucketAggregationWithAccessor,
+        force_flush: bool,
```

Comment:
Why is this argument required?

Reply:
There's a cache of 256 in each (sub)tree, that handles the docs when it is filled in one block.
After all docs have been collected the caches need to be flushed.
The parameter is to pass down the flush operation to the sub_trees. Each bucket can have a subtree (recursion)